scene/2d/physics_body_2d.cpp

```diff
@@ -1049,134 +1052,234 @@ void RigidBody2D::_reload_physics_characteristics() {
+					can_apply_constant_speed = false;
+					Vector2 motion_slide_norm = result.remainder.slide(result.collision_normal).normalized();
+					if (!motion_slide_norm.is_equal_approx(Vector2())) {
+						motion = motion_slide_norm * (motion_slide_up.length() - result.motion.slide(up_direction).length() - last_travel.slide(up_direction).length());
+					}
+				} else if ((sliding_enabled || !on_floor) && !(on_ceiling && !slide_on_ceiling && linear_velocity.dot(up_direction) > 0)) {
+					Vector2 slide_motion = result.remainder.slide(result.collision_normal);
+					if (slide_motion.dot(linear_velocity) > 0.0) {
+						motion = slide_motion;
+					} else {
+						motion = Vector2();
+					}
+					if (slide_on_ceiling && on_ceiling) {
+						if (linear_velocity.dot(up_direction) > 0) {
+							linear_velocity = linear_velocity.slide(result.collision_normal);
+						} else {
+							// Avoid acceleration in slope when falling.
+							linear_velocity = up_direction * up_direction.dot(linear_velocity);
+						}
+					}
 				} else {
 					motion = result.remainder;
+					if (on_ceiling && !slide_on_ceiling && linear_velocity.dot(up_direction) > 0) {
+						linear_velocity = linear_velocity.slide(up_direction);
+						motion = motion.slide(up_direction);
+					}
 				}
-			}
 
-			sliding_enabled = true;
+				last_travel = result.motion;
+			}
+			// Apply constant speed when the slope is downward.
+			else if (i == 0 && constant_speed_on_floor && first_slide && was_on_floor && !Math::is_equal_approx(floor_snap_length, 0) && _on_floor_if_snapped(was_on_floor)) {
+				can_apply_constant_speed = false;
+				sliding_enabled = true;
+				Transform2D gt = get_global_transform();
+				gt.elements[2] = prev_position;
+				set_global_transform(gt);
+
+				Vector2 motion_slide_norm = motion.slide(prev_floor_normal).normalized();
+				if (!motion_slide_norm.is_equal_approx(Vector2())) {
+					motion = motion_slide_norm * (motion_slide_up.length());
+					found_collision = true;
+				}
+			}
 		}
+		can_apply_constant_speed = !can_apply_constant_speed && !sliding_enabled;
+		sliding_enabled = true;
+		first_slide = false;
 
-		if (!found_collision || motion == Vector2()) {
+		if (!found_collision || motion.is_equal_approx(Vector2())) {
 			break;
 		}
 	}
 
+	_snap_on_floor(was_on_floor);
+
 	if (!on_floor && !on_wall) {
 		// Add last platform velocity when just left a moving platform.
-		linear_velocity += current_floor_velocity;
+		linear_velocity += current_platform_velocity;
 	}
 
-	if (!was_on_floor || snap == Vector2()) {
+	if (on_floor && linear_velocity.dot(up_direction) <= 0) {
+		linear_velocity = linear_velocity.slide(up_direction);
+	}
+
+	if (motion_results.size() == 0) {
+		return Ref<KinematicCollision2D>();
+	} else {
+		_set_motion_cache(motion_results.get(motion_results.size() - 1));
+		return motion_cache;
+	}
```

Comment:
The main advantages are: 

- this matches the api of `move_and_collide`
- in vast majority of case you only want the latest collision
- you can use the info to test if collided

At the price of an allocation.

If we don't keep this, it will be maybe worth adding a method `has_collided`.

What should I do?

Reply:
I would rather go for the extra method `has_collided` to avoid any possible confusion with `move_and_collide`, which works in a different way and has a different API. It seems clearer to provide just one way to get the collision information.